File: a2a-core-ts/src/task-store.ts
import { Task, TaskStatus, TaskState, Message, Artifact } from "./types";

const VALID_TRANSITIONS: Record<TaskStatus, TaskStatus[]> = {
  submitted: ["working", "canceled", "failed"],
  working: ["completed", "failed", "canceled"],
  completed: [],
  failed: [],
  canceled: [],
};

export class TaskStore {
  private tasks: Map<string, Task> = new Map();

  create(id: string, sessionId?: string, metadata?: Record<string, unknown>): Task {
    if (this.tasks.has(id)) {
      throw new Error(`Task ${id} already exists`);
    }

    const task: Task = {
      id,
      sessionId,
      status: {
        status: "submitted",
        timestamp: new Date().toISOString(),
      },
      messages: [],
      artifacts: [],
      metadata,
    };

    this.tasks.set(id, task);
    return task;
  }

  get(id: string): Task | undefined {
    return this.tasks.get(id);
  }

  updateStatus(id: string, newStatus: TaskStatus, message?: Message): Task {
    const task = this.tasks.get(id);
    if (!task) {
      throw new Error(`Task ${id} not found`);
    }

    const currentStatus = task.status.status;
    const allowed = VALID_TRANSITIONS[currentStatus];
    if (!allowed.includes(newStatus)) {
      throw new Error(
        `Invalid state transition: ${currentStatus} → ${newStatus}. Allowed: ${allowed.join(", ")}`
      );
    }

    task.status = {
      status: newStatus,
      message,
      timestamp: new Date().toISOString(),
    };

    return task;
  }

  addMessage(id: string, message: Message): Task {
    const task = this.tasks.get(id);
    if (!task) {
      throw new Error(`Task ${id} not found`);
    }
    task.messages.push(message);
    return task;
  }

  addArtifact(id: string, artifact: Artifact): Task {
    const task = this.tasks.get(id);
    if (!task) {
      throw new Error(`Task ${id} not found`);
    }
    artifact.index = task.artifacts.length;
    task.artifacts.push(artifact);
    return task;
  }

  cancel(id: string): Task {
    return this.updateStatus(id, "canceled");
  }

  list(): Task[] {
    return Array.from(this.tasks.values());
  }

  delete(id: string): boolean {
    return this.tasks.delete(id);
  }
}

File: a2a-core-py/a2a_core/task_store.py
"""Thread-safe in-memory task store."""

from __future__ import annotations

import threading
from datetime import datetime, timezone

from .models import Artifact, Message, Task, TaskState, TaskStatus


VALID_TRANSITIONS: dict[TaskStatus, list[TaskStatus]] = {
    TaskStatus.SUBMITTED: [TaskStatus.WORKING, TaskStatus.CANCELED, TaskStatus.FAILED],
    TaskStatus.WORKING: [TaskStatus.COMPLETED, TaskStatus.FAILED, TaskStatus.CANCELED],
    TaskStatus.COMPLETED: [],
    TaskStatus.FAILED: [],
    TaskStatus.CANCELED: [],
}


class TaskStore:
    def __init__(self) -> None:
        self._tasks: dict[str, Task] = {}
        self._lock = threading.Lock()

    def _now(self) -> str:
        return datetime.now(timezone.utc).isoformat()

    def create(
        self,
        task_id: str,
        session_id: str | None = None,
        metadata: dict | None = None,
    ) -> Task:
        with self._lock:
            if task_id in self._tasks:
                raise ValueError(f"Task {task_id} already exists")
            task = Task(
                id=task_id,
                sessionId=session_id,
                status=TaskState(status=TaskStatus.SUBMITTED, timestamp=self._now()),
                messages=[],
                artifacts=[],
                metadata=metadata,
            )
            self._tasks[task_id] = task
            return task

    def get(self, task_id: str) -> Task | None:
        with self._lock:
            return self._tasks.get(task_id)

    def update_status(
        self,
        task_id: str,
        new_status: TaskStatus,
        message: Message | None = None,
    ) -> Task:
        with self._lock:
            task = self._tasks.get(task_id)
            if task is None:
                raise ValueError(f"Task {task_id} not found")

            current = task.status.status
            allowed = VALID_TRANSITIONS[current]
            if new_status not in allowed:
                raise ValueError(
                    f"Invalid state transition: {current.value} → {new_status.value}. "
                    f"Allowed: {[s.value for s in allowed]}"
                )

            task.status = TaskState(
                status=new_status,
                message=message,
                timestamp=self._now(),
            )
            return task

    def add_message(self, task_id: str, message: Message) -> Task:
        with self._lock:
            task = self._tasks.get(task_id)
            if task is None:
                raise ValueError(f"Task {task_id} not found")
            task.messages.append(message)
            return task

    def add_artifact(self, task_id: str, artifact: Artifact) -> Task:
        with self._lock:
            task = self._tasks.get(task_id)
            if task is None:
                raise ValueError(f"Task {task_id} not found")
            artifact.index = len(task.artifacts)
            task.artifacts.append(artifact)
            return task

    def cancel(self, task_id: str) -> Task:
        return self.update_status(task_id, TaskStatus.CANCELED)

    def list_tasks(self) -> list[Task]:
        with self._lock:
            return list(self._tasks.values())

    def delete(self, task_id: str) -> bool:
        with self._lock:
            return self._tasks.pop(task_id, None) is not None
